### PR TITLE
Generate Prisms and Lenses for ps types

### PR DESCRIPTION
**Pull request by boothead**
At the moment I'm generating prisms for every constructor, but lenses only for types with one constructor that contains a record.

The example code in https://github.com/eskimor/servant-purescript compiles with this (separate PR to come for those changes)

Still to do:

- [x] Add a note about purescript-profunctor-lenses being required

### CODE REVIEW

File: test/Spec.hs [hunk 1]
```diff
@@ -76,6 +79,102 @@ allTests =
                           , ""
                           , "derive instance genericBar :: (Generic a, Generic b, Generic (m b)) => Generic (Bar a b m c)"
                           , ""
+                          , "--------------------------------------------------------------------------------"
+                          , "_Bar1 :: PrismP (Bar a b m c) (Maybe a)"
```
Comment (eskimor) [inline test/Spec.hs]: Does this code compile? Isn't there a forall a b m c missing?

Comment (boothead) [inline test/Spec.hs]: Ah you're right - didn't notice as I don't have any Bar like types in my code (thankfully) :-) I'll push a fix.

**Comment by boothead** [comment]
One last problem is that we now get 3 warning in generated code about unspecified imports. If that's acceptable then I think this is good.

File: src/Language/PureScript/Bridge/Printer.hs [hunk 2]
```diff
@@ -58,21 +59,36 @@ moduleToText m = T.unlines $
...
   : "module " <> psModuleName m <> " where\n"
   : map importLineToText (Map.elems (psImportLines m))
-  ++ [ "\nimport Data.Generic (class Generic)\n\n" ]
+  ++ [ ""
+     , "import Prelude"
+     , "import Data.Generic (class Generic)"
+     , "import Data.Maybe"
+     , "import Data.Lens"
+     , ""
+     ]
```
Comment (eskimor) [inline src/Language/PureScript/Bridge/Printer.hs]: ideally we would add all those imports to the generated module psImportLines, this way also duplicate imports would be avoided. I can't remember why I did not do this back then, but now that there are more imports lines added, it seems to be a good idea.

Comment (eskimor) [inline src/Language/PureScript/Bridge/Printer.hs]: If the generated code does not generate warnings, this would definitely be better.

Comment (boothead) [inline src/Language/PureScript/Bridge/Printer.hs]: I tried that, but it only generates lines like `import Data.Maybe ()` I think the warnings are fixable

**Comment by boothead** [comment]
I've noticed one more bug in Prism generation that I'll need to fix. I'll do that shortly

File: src/Language/PureScript/Bridge/Printer.hs (hunk 2 above)
Comment (eskimor) [inline src/Language/PureScript/Bridge/Printer.hs]: Yeah - you need to specify what you are going to import. All imports are explicit, this then also fixes the warnings about unspecified imports.

Comment (eskimor) [inline src/Language/PureScript/Bridge/Printer.hs]: If you add them statically and a used type also depends on Maybe for example we get warnings about redundant imports, so adding them to psImportLines is the correct approach.

**Comment by eskimor** [comment]
Please drop the stack.yaml file, so the travis build can succeed. Thanks!

**Comment by boothead** [comment]
Oops, my bad. Fixed. I'll try and get some time to look at the imports this weekend too.

**Comment by boothead** [comment]
I've moved the imports into the same mechanism as adding import lines, but I think there will still be a few warnings (see last commit message)

**Comment by boothead** [comment]
Just ran this version on my large-ish project and it all still works.

**Comment by eskimor** [comment]
Nice work -thanks!

**Pull request merged**